src/Translation/GoogleTranslateApi.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+namespace App\Translation;
+
+use Google\Cloud\Core\Exception\ServiceException;
+use Google\Cloud\Translate\V2\TranslateClient;
+use Psr\Log\LoggerInterface;
+
+class GoogleTranslateApi implements TranslationApiInterface
+{
+  private const LONG_LANGUAGE_CODE = [
+    'zh-CN',
+    'zh-TW',
+  ];
+
+  private TranslateClient $client;
+  private LoggerInterface $logger;
+
+  public function __construct(TranslateClient $client, LoggerInterface $logger)
+  {
+    $this->client = $client;
+    $this->logger = $logger;
+  }
+
+  public function translate(string $text, ?string $source_language, string $target_language): ?TranslationResult
+  {
+    $target_language = $this->transformLanguageCode($target_language);
+    $source_language = $this->transformLanguageCode($source_language);
+
+    try {
+      $result = $this->client->translate($text, [
+        'source' => $source_language,
+        'target' => $target_language,
+        'format' => 'text',
+      ]);
+    } catch (ServiceException $e) {
+      $this->logger->error("Google translate client exception, source: {$source_language}, target: {$target_language}, text: {$text}, message: {$e->getMessage()}");
+
+      return null;
+    }
+
+    if (null === $result) {
+      return null;
+    }
+
+    $translation_result = new TranslationResult();
+    $translation_result->provider = 'google';
+    if (null == $source_language) {
```

Comment:
Should this be `===`?

Reply:
there are a few values other than `null` that can technically evaluate to true here. But i think it's ok because we assume `source_language` is a valid language code or `null`, so it can't be empty string, empty array, 0 or `false`